vllm/core/block_manager_v2.py

```diff
@@ -116,18 +116,39 @@ def allocate(self, seq_group: SequenceGroup) -> None:
         for seq in waiting_seqs[1:]:
             self.block_tables[seq.seq_id] = block_table.fork()
 
-    def can_append_slot(self, seq_group: SequenceGroup) -> bool:
-        # Simple heuristic: If there is at least one free block
-        # for each sequence, we can append.
+    def can_append_slots(self, seq_group: SequenceGroup,
+                         num_lookahead_slots: int) -> bool:
+        """Determine if there is enough space in the GPU KV cache to continue
+        generation of the specified sequence group.
+
+        We use a worst-case heuristic: assume each touched block will require a
+        new allocation (either via CoW or new block). We can append slots if the
+        number of touched blocks is less than the number of free blocks.
+
+        "Lookahead slots" are slots that are allocated in addition to the slots
+        for known tokens. This is used by speculative decoding when speculating
+        future tokens.
+        """
+        num_touched_blocks = 0
+        for seq in seq_group.get_seqs(status=SequenceStatus.RUNNING):
+            block_table = self.block_tables[seq.seq_id]
+            num_new_tokens = seq.get_len() - block_table.num_full_slots
```

Comment:
what's the definition of `num_new_tokens` here?

Reply:
This is the number of tokens that do not have a slot allocated in the block table. Will add a comment.

Reply:
From my understanding, the `num_new_tokens == len(unseen_token_ids)`?

Reply:
yes. I will rename them so it's consistent.